### PR TITLE
agentless: Multi-cluster tproxy

### PR DESCRIPTION
**Pull request by ishustava**
Changes proposed in this PR:
- Redirect traffic to the consul-dataplane's DNS proxy when DNS redirection is enabled. This means we no longer need to configure recursors for the clients and servers
- To fall back to kube dns when consul cannot resolve names, we define our own DNSConfig on the pod that uses Consul first, but falls back to the config from `/etc/resolv.conf` in the cluster.

How I've tested this PR:
acceptance tests

How I expect reviewers to test this PR:
👀 

Checklist:
- [x] Tests added
- [ ] CHANGELOG entry added 
  > HashiCorp engineers only, community PRs should not add a changelog entry.
  > Entries should use present tense (e.g. Add support for...)

### CODE REVIEW

File: charts/consul/templates/_helpers.tpl [hunk 1]
```diff
@@ -148,19 +148,6 @@ is passed to consul as a -config-file param on command line.
...
-to the consul agent command.
-*/}}
-{{- define "consul.recursors" -}}
-              recursor_flags=""
-              for ip in $(cat /etc/resolv.conf | grep nameserver | cut -d' ' -f2)
-              do
-                 recursor_flags="$recursor_flags -recursor=$ip"
-              done
-{{- end -}}
-
```
Comment (thisisnotashwin) [inline charts/consul/templates/_helpers.tpl:163]: YAYYYYY!!!!

File: control-plane/connect-inject/peering_acceptor_controller_test.go [hunk 2]
```diff
@@ -1194,7 +1194,7 @@ func TestAcceptorUpdateStatusError(t *testing.T) {
 			reconcileErr: errors.New("this is an error"),
 			expStatus: v1alpha1.PeeringAcceptorStatus{
 				Conditions: v1alpha1.Conditions{
-					{
+					v1alpha1.Condition{
```
[the remark this answers is not in the document]
Comment (thisisnotashwin) [reply control-plane/connect-inject/peering_acceptor_controller_test.go:1135]: This is so weird. I remember getting linter failures when they initially existed and had to delete them. 🤔

File: control-plane/connect-inject/dns.go [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+	// These defaults are taken from the /etc/resolv.conf man page
+	// and are used by the dns library.
+	defaultDNSOptionNdots    = 1
+	defaultDNSOptionTimeout  = 5
+	defaultDNSOptionAttempts = 2
+)
+
+func (w *MeshWebhook) configureDNS(pod *corev1.Pod, k8sNS string) error {
+	// First, we need to determine the nameservers configured in this cluster from /etc/resolv.conf.
+	etcResolvConf := "/etc/resolv.conf"
```
Comment (thisisnotashwin) [inline control-plane/connect-inject/dns.go:22]: We should make this a constant. WDYT?

File: control-plane/connect-inject/dns.go [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+
+	// Set DNS policy on the pod to None because we want DNS to work according to the config we will provide.
+	pod.Spec.DNSPolicy = corev1.DNSNone
+
+	// Set the consul-dataplane's DNS server as the first server in the list (i.e. localhost).
+	// We want to do that so that when consul cannot resolve the record, we will fall back to the nameservers
+	// configured in our /etc/resolv.conf. It's important to add Consul DNS as the first nameserver because
+	// if we put kube DNS first, it will return NXDOMAIN response and a DNS client will not fall back to other nameservers.
+	if pod.Spec.DNSConfig == nil {
+		consulDPAddress := "127.0.0.1"
```
Comment (thisisnotashwin) [inline control-plane/connect-inject/dns.go:39]: potentially make this a constant as well

File: control-plane/connect-inject/dns.go [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+			options = append(options, corev1.PodDNSConfigOption{
+				Name:  "attempts",
+				Value: pointer.String(strconv.Itoa(cfg.Attempts)),
+			})
+		}
+
+		// Replace release namespace in the searches with the pod namespace.
+		// This is so that the searches we generate will be for the pod's namespace
+		// instead of the namespace of the connect-injector. E.g. instead of
+		// consul.svc.cluster.local it should be <pod ns>.svc.cluster.local.
```
Comment (thisisnotashwin) [inline control-plane/connect-inject/dns.go:66]: Really appreciate the comments 🙏

**Review by thisisnotashwin — APPROVED** [review]
Have a few aesthetic comments but this is excellent!! Your comments made reviewing and understanding what was happening super easy! Great job!

File: control-plane/connect-inject/peering_acceptor_controller_test.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (ishustava) [reply control-plane/connect-inject/peering_acceptor_controller_test.go:1135]: Yeah I'm confused as well

File: control-plane/connect-inject/peering_acceptor_controller_test.go [hunk 6]
```diff
@@ -620,9 +620,9 @@ func TestReconcile_CreateUpdatePeeringAcceptor(t *testing.T) {
 			decodedTokenData, err := base64.StdEncoding.DecodeString(string(createdSecret.Data["data"]))
 			require.NoError(t, err)
 
-			require.Contains(t, string(decodedTokenData), "\"CA\":null")
+			require.Contains(t, string(decodedTokenData), "\"CA\":")
```
[the remark this answers is not in the document]
Comment (ndhanushkodi) [reply control-plane/connect-inject/peering_acceptor_controller_test.go:567]: Ahh thanks for bumping that!!

File: control-plane/connect-inject/peering_dialer_controller_test.go [hunk 7]
```diff
@@ -424,9 +427,12 @@ func TestReconcile_VersionAnnotationPeeringDialer(t *testing.T) {
...
 			// Create test consul server.
 			acceptorPeerServer, err := testutil.NewTestServerConfigT(t, func(c *testutil.TestServerConfig) {
-				// We set the datacenter because the server name, typically formatted as "server.<datacenter>.<domain>"
+				// We set different cluster id for the connect CA because the server name,
+				// typically formatted as server.dc1.peering.<cluster_id>.consul
 				// must be unique on the acceptor and dialer peers.
-				c.Datacenter = "acceptor-dc"
+				c.Connect["ca_config"] = map[string]interface{}{
+					"cluster_id": "00000000-2222-3333-4444-555555555555",
+				}
```
Comment (ndhanushkodi) [inline control-plane/connect-inject/peering_dialer_controller_test.go:435]: ❤️

File: acceptance/tests/peering/peering_connect_test.go [hunk 8]
```diff
@@ -31,10 +31,6 @@ func TestPeering_Connect(t *testing.T) {
 		t.Skipf("skipping this test because peering is not supported in version %v", cfg.ConsulVersion.String())
 	}
 
-	if cfg.EnableTransparentProxy {
-		t.Skipf("skipping because no t-proxy support")
-	}
-
```
Comment (ndhanushkodi) [inline acceptance/tests/peering/peering_connect_test.go:37]: omg so this works???

Comment (ishustava) [reply acceptance/tests/peering/peering_connect_test.go:37]: yup! it works! 🎉

**Review by curtbushko — APPROVED** [review]

**Pull request merged**